File: datastructures/graph/Graph.java
import java.util.ArrayList;
import java.util.List;


public class Graph {
    private int V;
    private int E;
    private List<Integer>[] adj;

    public Graph(int V) {
        this.V = V;
        this.E = 0;
        adj = (List<Integer>[]) new ArrayList[V];
        //adj = new ArrayList<Integer>[V];
        for (int v = 0; v < V; v++) {
            adj[v] = new ArrayList<>();
        }
    }

    public int V() {
        return V;
    }

    public int E() {
        return E;
    }

    public void addEdge(int v, int w) {
        adj[v].add(w);
//        adj[w].add(v);
        E++;
    }

    public Iterable<Integer> adj(int v) {
        return adj[v];
    }

    public String toString() {
        String s = V + " vertices, " + E + " edges\n";
        for (int v = 0; v < V; v++) {
            s += v + ": ";
            for (int w : this.adj(v)) {
                s += w + " ";
            }
            s += "\n";
        }
        return s;
    }

    public void testConnectedness(int w) {
        DFS search = new DFS(this, w);
        for (int v = 0; v < V(); v++) {
            if (search.marked(v)) {
                System.out.print(v + " ");
            }
        }
        System.out.println();

        if (search.count() != V()) {
            System.out.print("NOT ");
        }
        System.out.println("connected");
    }

    public static void main(String[] args) {
        // V E
        Graph g = new Graph(6);
        g.addEdge(0, 5);
        g.addEdge(2, 4);
        g.addEdge(2, 3);
        g.addEdge(1, 2);
        g.addEdge(0, 1);
        g.addEdge(3, 4);
        g.addEdge(3, 5);
        g.addEdge(0, 2);

        System.out.println(g);
        int s = 0;
        System.out.println("Testing connectedness for v: " + s);
        g.testConnectedness(s);
        s = 3;
        System.out.println("Testing connectedness for v: " + s);
        g.testConnectedness(s);

        s = 0;
        System.out.println("Testing paths to " + s);
        Paths path = new DFSPathFinder(g, s);
        for (int v = 0; v < g.V(); v++) {
            if (path.hasPathTo(v)) {
                System.out.print(s + " to " + v + ": ");
                for (int x : path.pathTo(v)) {
                    if (x == s) System.out.print(x);
                    else        System.out.print(x + "-");
                }
                System.out.println();
            }
        }
    }
}

File: datastructures/graph/graph.js
"use strict";

const DFS = require('./dfs');

class Graph {
    constructor(V) {
        this.V = V;
        this.E = 0;
        this.adj = [];
        for (let i = 0; i < V; i++) {
            this.adj.push([]);
        }
    }

    addEdge(v, w) {
        this.adj[v].push(w);
        this.adj[w].push(v);
        this.E += 1;
    }

    toString() {
        let s = `Vertices: ${this.V}, Edges: ${this.E}\n`;
        for (const i in this.adj) {
            s += `${i}: `;
            for (const w of this.adj[i]) {
                s += `${w} `;
            }
            s += `\n`;
        }
        return s;
    }
    
    testConnectedness(s) {
        const dfs = new DFS(this, s);
        let o = `${s}: `;
        for (const i in dfs.marked) {
            if (dfs.marked[i]) {
                o += `${i} `;
            }
        }
        o += "\n";
        if (dfs.count != this.V) {
            o += "NOT ";
        }
        o += "connected\n";
        console.log(o);
    }
}

if (require.main === module) {

    const g = new Graph(13);
    g.addEdge(0, 5);
    g.addEdge(4, 3);
    g.addEdge(0, 1);
    g.addEdge(9, 12);
    g.addEdge(6, 4);
    g.addEdge(5, 4);
    g.addEdge(0, 2);
    g.addEdge(11, 12);
    g.addEdge(9, 10);
    g.addEdge(0, 6);
    g.addEdge(7, 8);
    g.addEdge(9, 11);
    g.addEdge(5, 3);

    console.log(g.toString());
    console.log(`Testing connectedness for v: 0`);
    g.testConnectedness(0);
    console.log(`Testing connectedness for v: 9`);
    g.testConnectedness(9);    
}
